pkg/alert/alert.go

```diff
@@ -0,0 +1,427 @@
+							"_source": {
+								"includes": [
+									"name",
+									"startTime",
+									"pre.clusterDomain",
+									"tm.landscape",
+									"tm.tr.id",
+									"tm.cloudprovider",
+									"tm.k8s_version",
+									"tm.operating_system"
+								]
+							},
+							"size": 1
+						}
+					}
+				}
+			}
+		}
+	}`, alert.cfg.EvalTimeDays, alert.cfg.ContinuousFailureThreshold, alert.cfg.ContinuousFailureThreshold)
+
+	var testContextAggregation testContextAggregation
+	alert.elasticRequest(`/testmachinery-*/_search`, "GET", payloadFormated, &testContextAggregation)
+	alert.log.Info(fmt.Sprintf("retrieved %d distinct test aggregations", len(testContextAggregation.Aggs.TestContext.TestDetailsRaw)))
+
+	return testContextAggregation
+}
+
+//extractTestDetailItems parses raw elasticsearch test aggregations into test details
+func extractTestDetailItems(testContextAggregation testContextAggregation) map[string]TestDetails {
+	contextToTestDetailMap := make(map[string]TestDetails)
+	for _, testDoc := range testContextAggregation.Aggs.TestContext.TestDetailsRaw {
+		testDocDetails := testDoc.Details.Hits.Hits[0].Source
+		testFailedContinuously := true
+		for _, trendItem := range testDoc.SuccessTrend.Hits.Hits {
+			if trendItem.Source.Pre.PhaseNum != 0 {
+				testFailedContinuously = false
+				break
+			}
+		}
+		parsedTestDetail := TestDetails{
+			Name:                testDocDetails.Name,
+			LastFailedTimestamp: testDocDetails.LastTimestamp,
+			OperatingSystem:     testDocDetails.TM.OperatingSystem,
+			K8sVersion:          testDocDetails.TM.K8sVersion,
+			Cloudprovider:       testDocDetails.TM.Cloudprovider,
+			Landscape:           testDocDetails.TM.Landscape,
+			SuccessRate:         testDoc.SuccessRate.Value,
+			Context:             testDoc.Testcontext,
+			FailedContinuously:  testFailedContinuously,
+		}
+		contextToTestDetailMap[testDoc.Testcontext] = parsedTestDetail
+	}
+	return contextToTestDetailMap
+}
+
+//removeSuccessfulTests removes tests from input map, that are not exceeding given thresholds and are therefore considered successful
+func (alert *Alert) removeSuccessfulTests(contextToTestDetailMap map[string]TestDetails) map[string]TestDetails {
+	testsSizeBefore := len(contextToTestDetailMap)
+	for key, value := range contextToTestDetailMap {
+		if !value.FailedContinuously && value.SuccessRate >= float64(alert.cfg.SuccessRateThresholdPercent) {
+			delete(contextToTestDetailMap, key)
+		}
+	}
+	alert.log.Info(fmt.Sprintf("removed %d/%d tests, because they are successful", testsSizeBefore-len(contextToTestDetailMap), testsSizeBefore))
+	return contextToTestDetailMap
+}
+
+//removeAlreadyFiledAlerts removes tests based on given test exclude regexp patterns
+func (alert *Alert) removeExcludedTests(tests map[string]TestDetails) {
+	testsSizeBefore := len(tests)
+	for _, patternStr := range alert.cfg.TestsToExclude {
+		alert.log.Info(fmt.Sprintf("filtering out tests with expression %s", patternStr))
```

Comment:
maybe use logging levels so logging can be increased if needed `alert.log.V(3).Info()`